radix-engine-common/src/types/consensus.rs

```diff
@@ -0,0 +1,89 @@
+use sbor::Sbor;
+
+/// A type-safe consensus epoch number.
+/// Assuming one epoch per minute (i.e. much faster progression than designed), this gives us time
+/// until A.D. ~5700.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Sbor)]
+#[sbor(transparent)]
+pub struct Epoch(u32);
+
+/// An index of a specific validator within the current validator set.
+/// To be exact: a `ValidatorIndex` equal to `k` references the `k-th` element returned by the
+/// iterator of the `IndexMap<ComponentAddress, Validator>` in this epoch's active validator set
+/// (which is expected to be sorted by stake, descending).
+/// This uniquely identifies the validator, while being shorter than `ComponentAddress` (we do care
+/// about the constant factor of the space taken by `LeaderProposalHistory` under prolonged liveness
+/// break scenarios).
+pub type ValidatorIndex = u8;
+
+impl Epoch {
+    /// Creates a zero epoch (i.e. pre-genesis).
+    pub fn zero() -> Self {
+        Self::of(0)
+    }
+
+    /// Creates an epoch of the given number.
+    pub fn of(number: u32) -> Self {
+        Self(number)
+    }
+
+    /// Returns a raw epoch number.
+    pub fn number(&self) -> u32 {
+        self.0
+    }
+
+    /// Creates an epoch immediately following this one.
+    pub fn next(&self) -> Self {
+        self.after(1)
+    }
+
+    /// Creates an epoch following this one after the given number of epochs.
+    pub fn after(&self, epoch_count: u32) -> Self {
+        self.relative(epoch_count as i64)
+    }
+
+    /// Creates an epoch immediately preceding this one.
+    pub fn previous(&self) -> Self {
+        self.relative(-1)
+    }
+
+    /// Creates an epoch of a number relative to this one.
+    fn relative(&self, epoch_count: i64) -> Self {
+        Self(u32::try_from(self.0 as i64 + epoch_count).unwrap())
```

Comment:
A few things:
* It feels scary this panics if it overflows or underflows
* Why an i64? Rather than, say, an i32?
* Can we make this return an Option<Self> instead?

Reply:
> It feels scary this panics if it overflows or underflows. Can we make this return an Option instead?

AFAIU, the previous production behavior (i.e. `next_epoch = current_epoch + 1`) was not panicking, but silently wrapping back to 0. Is it a better behavior?

I mean, the code must do _something_ when it reaches those values. Returning an `Option` only changes _which_ code must do that.

My take is: we don't expect billions of epochs to be reached during our lifetimes. So if it happens, it is either because of a bug or a hack, and I'd appreciate if the system failed (with some obvious indication of the reason: a message or stack trace).

Let me explain this in rustdocs.

> Why an i64? Rather than, say, an i32?

This has a good reason actually:
A difference between two `u32`s can be in range `[-u32::MAX, u32::MAX]`.
The `i32` cannot represent that range (observe that `i32::MAX = u32::MAX / 2`), so `i64` is needed. (technically, `i33` could be fine, if it existed)